### PR TITLE
Fix handling of non_existing entries in TFRecord reader

### PR DESCRIPTION
**Pull request by JanuszL**
- DALI assumes that all features provided in the TFRecord reader
  exists in every entry, otherwise protobuf will throw.
  This PR lifts that assumption makes the reader return
  empty tensor when the desired entry has not been found.
  It should be in pair what TensorFlow reader does.

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It fixes handling of non_existing entries in TFRecord reader

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     DALI assumes that all features provided in the TFRecord reader exists in every entry, otherwise protobuf will throw.
     This PR lifts that assumption makes the reader return empty tensor when the desired entry has not been found.
     It should be in pair what TensorFlow reader does.
 - Affected modules and functionalities:
     tfrecord_parser.h
 - Key points relevant for the review:
     should we return empty tensor and/or warn/enforce that all entries exits
 - Validation and testing:
     test is added
 - Documentation (including examples):
     TFRecord Reader documentation is updated

**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/parser/tfrecord_parser.h [hunk 1]
```diff
@@ -66,7 +66,26 @@ class TFRecordParser : public Parser<Tensor<CPUBackend>> {
...
+        switch (f.GetType()) {
+          case FeatureType::int64:
+            (void) output.mutable_data<int64_t>();
+            break;
+          case FeatureType::string:
+            (void) output.mutable_data<uint8_t>();
+            break;
+          case FeatureType::float32:
+            (void) output.mutable_data<float>();
+            break;
```
Comment (mzient) [inline dali/operators/reader/parser/tfrecord_parser.h:83]: Setting the type as a side effect of calling `mutable_data` is ugly.
```suggestion
            output.Resize({0}, TypeTable::GetTypeInfo(DALI_INT64));
            break;
          case FeatureType::string:
            output.Resize({0}, TypeTable::GetTypeInfo(DALI_UINT8));
            break;
          case FeatureType::float32:
            output.Resize({0}, TypeTable::GetTypeInfo(DALI_FLOAT));
            break;
```

File: dali/test/python/test_operator_readers_index.py [hunk 2]
```diff
@@ -146,3 +146,25 @@ def test_tfrecord_reader_alias():
...
+def test_tfrecord_reader_alias():
+    tfrecord = os.path.join(get_dali_extra_path(), 'db', 'tfrecord', 'train')
+    tfrecord_idx = os.path.join(get_dali_extra_path(), 'db', 'tfrecord', 'train.idx')
+    pipe = tfrecord_pipe_empty_fields(tfrecord, tfrecord_idx)
+    pipe.build()
+    out = pipe.run()
+    for tensor in out[0]:
+        assert len(np.array(tensor)) != 0
+    for tensor in out[1]:
+        assert len(np.array(tensor)) == 0
```
Comment (mzient) [inline dali/test/python/test_operator_readers_index.py:168]:
```suggestion
        assert len(np.array(tensor)) == 0
        assert np.dtype == np.int64
```

File: dali/test/python/test_operator_readers_index.py [hunk 3]
```diff
@@ -146,3 +146,25 @@ def test_tfrecord_reader_alias():
...
+    for tensor in out[2]:
+        assert len(np.array(tensor)) == 0
```
Comment (mzient) [inline dali/test/python/test_operator_readers_index.py:170]:
```suggestion
        assert len(np.array(tensor)) == 0
        assert np.dtype == np.uint8        
```

Comment (JanuszL) [reply dali/test/python/test_operator_readers_index.py]: Done

File: dali/test/python/test_operator_readers_index.py (hunk 2 above)
Comment (JanuszL) [reply dali/test/python/test_operator_readers_index.py]: Done

File: dali/operators/reader/parser/tfrecord_parser.h (hunk 1 above)
Comment (JanuszL) [reply dali/operators/reader/parser/tfrecord_parser.h]: Done

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by JanuszL** [comment]
@Riyria-was-taken - this should fix the issue with the TFRecird you mentioned some time ago.

**Pull request merged**